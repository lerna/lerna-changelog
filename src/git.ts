const execa = require("execa");

export function getRootPath() {
  const cwd = process.cwd();
  return execa.sync("git", ["rev-parse", "--show-toplevel"], { cwd }).stdout;
}

export async function changedPaths(sha: string): Promise<string[]> {
  const result = await execa("git", ["show", "-m", "--name-only", "--pretty=format:", "--first-parent", sha]);
  return result.stdout.split("\n");
}

/**
 * All existing tags in the repository
 */
export function listTagNames(): string[] {
  return execa
    .sync("git", ["tag"])
    .stdout.split("\n")
    .filter(Boolean);
}

/**
 * The latest reachable tag starting from HEAD
 */
export function lastTag(): string {
  return execa.sync("git", ["describe", "--abbrev=0", "--tags"]).stdout;
}

export interface CommitListItem {
  sha: string;
  refName: string;
  summary: string;
  date: string;
}

export function parseLogMessage(commit: string): CommitListItem | null {
  const parts = commit.match(/hash<(.+)> ref<(.*)> message<(.*)> date<(.*)>/) || [];

  if (!parts || parts.length === 0) {
    return null;
  }

  return {
    sha: parts[1],
    refName: parts[2],
    summary: parts[3],
    date: parts[4],
  };
}

export function listCommits(from: string, to: string = ""): CommitListItem[] {
  // Prints "hash<short-hash> ref<ref-name> message<summary> date<date>"
  // This format is used in `getCommitInfos` for easily analize the commit.
  return execa
    .sync("git", [
      "log",
      "--oneline",
      "--pretty=hash<%h> ref<%D> message<%s> date<%cd>",
      "--date=short",
      `${from}..${to}`,
    ])
    .stdout.split("\n")
    .filter(Boolean)
    .map(parseLogMessage)
    .filter(Boolean);
}
