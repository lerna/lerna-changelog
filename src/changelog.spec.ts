jest.mock("../src/progress-bar");
jest.mock("../src/changelog");
jest.mock("../src/git-hosting-api/github-api");
jest.mock("./git");
jest.mock("./fetch");

describe("Changelog", () => {
  describe("packageFromPath", () => {
    const MockedChangelog = require("./changelog").default;

    const TESTS = [
      ["", ""],
      ["foo.js", ""],
      ["packages/foo.js", ""],
      ["packages/foo/bar.js", "foo"],
      ["packages/foo/bar/baz.js", "foo"],
      ["packages/@foo/bar.js", "@foo"],
      ["packages/@foo/bar/baz.js", "@foo/bar"],
    ];

    for (let [input, expected] of TESTS) {
      it(`${input} -> ${expected}`, () => {
        const changelog = new MockedChangelog();
        expect(changelog.packageFromPath(input)).toEqual(expected);
      });
    }
  });

  describe("getCommitInfos", () => {
    beforeEach(() => {
      require("./fetch").__resetMockResponses();

      require("./git").listCommits.mockImplementation(() => [
        {
          sha: "a0000005",
          refName: "HEAD -> master, tag: v0.2.0, origin/master, origin/HEAD",
          summary: "chore(release): releasing component",
          date: "2017-01-01",
        },
        {
          sha: "a0000004",
          refName: "",
          summary: "Merge pull request #2 from my-feature",
          date: "2017-01-01",
        },
        {
          sha: "a0000003",
          refName: "",
          summary: "feat(module) Add new module (#2)",
          date: "2017-01-01",
        },
        {
          sha: "a0000002",
          refName: "",
          summary: "refactor(module) Simplify implementation",
          date: "2017-01-01",
        },
        {
          sha: "a0000001",
          refName: "tag: v0.1.0",
          summary: "chore(release): releasing component",
          date: "2017-01-01",
        },
      ]);

      require("./git").listTagNames.mockImplementation(() => ["v0.2.0", "v0.1.1", "v0.1.0", "v0.0.1"]);

      require("./git").changedPaths.mockImplementation(() => []);

      const usersCache = {
        "https://api.github.com/users/test-user": {
          body: {
            login: "test-user",
            html_url: "https://github.com/test-user",
            name: "Test User",
          },
        },
      };
      const issuesCache = {
        "https://api.github.com/repos/lerna/lerna-changelog/issues/2": {
          body: {
            number: 2,
            title: "This is the commit title for the issue (#2)",
            labels: [{ name: "Type: New Feature" }, { name: "Status: In Progress" }],
            user: usersCache["https://api.github.com/users/test-user"].body,
          },
        },
      };
      require("./fetch").__setMockResponses({
        ...usersCache,
        ...issuesCache,
      });
    });

    afterEach(() => {
      jest.resetAllMocks();
    });

    it("parse commits with different tags", async () => {
      const MockedChangelog = require("./changelog").default;
      const changelog = new MockedChangelog();
      const commitsInfo = await changelog.getCommitInfos();

      expect(commitsInfo).toMatchSnapshot();
    });
  });

  describe("getCommitters", () => {
    beforeEach(() => {
      require("./fetch").__resetMockResponses();

      const usersCache = {
        "https://api.github.com/users/test-user": {
          body: {
            login: "test-user",
            html_url: "https://github.com/test-user",
            name: "Test User",
          },
        },
        "https://api.github.com/users/test-user-1": {
          body: {
            login: "test-user-1",
            html_url: "https://github.com/test-user-1",
            name: "Test User 1",
          },
        },
        "https://api.github.com/users/test-user-2": {
          body: {
            login: "test-user-2",
            html_url: "https://github.com/test-user-2",
            name: "Test User 2",
          },
        },
        "https://api.github.com/users/user-bot": {
          body: {
            login: "user-bot",
            html_url: "https://github.com/user-bot",
            name: "User Bot",
          },
        },
      };
      require("./fetch").__setMockResponses(usersCache);
    });

    it("get list of valid commiters", async () => {
      const MockedChangelog = require("./changelog").default;
      const changelog = new MockedChangelog({
        ignoreCommitters: ["user-bot"],
      });

      const testCommits = [
        {
          commitSHA: "a0000004",
          gitHostingIssue: { user: { login: "test-user-1" } },
        },
        {
          commitSHA: "a0000003",
          gitHostingIssue: { user: { login: "test-user-2" } },
        },
        { commitSHA: "a0000002", gitHostingIssue: { user: { login: "user-bot" } } },
        { commitSHA: "a0000001" },
      ];
      const committers = await changelog.getCommitters(testCommits);

      expect(committers).toEqual([
        {
          login: "test-user-1",
          html_url: "https://github.com/test-user-1",
          name: "Test User 1",
        },
        {
          login: "test-user-2",
          html_url: "https://github.com/test-user-2",
          name: "Test User 2",
        },
      ]);
    });
  });
});
