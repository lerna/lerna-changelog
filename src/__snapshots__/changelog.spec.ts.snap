// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Changelog getCommitInfos parse commits with different tags 1`] = `
Array [
  Object {
    "commitSHA": "a0000005",
    "date": "2017-01-01",
    "issueNumber": null,
    "message": "chore(release): releasing component",
    "packages": Array [],
    "tags": Array [
      "v0.2.0",
    ],
  },
  Object {
    "categories": Array [
      ":rocket: New Feature",
    ],
    "commitSHA": "a0000004",
    "date": "2017-01-01",
    "gitHostingIssue": Object {
      "labels": Array [
        Object {
          "name": "Type: New Feature",
        },
        Object {
          "name": "Status: In Progress",
        },
      ],
      "number": 2,
      "title": "This is the commit title for the issue (#2)",
      "user": Object {
        "html_url": "https://github.com/test-user",
        "login": "test-user",
        "name": "Test User",
      },
    },
    "issueNumber": "2",
    "message": "Merge pull request #2 from my-feature",
    "packages": Array [],
    "tags": undefined,
  },
  Object {
    "categories": Array [
      ":rocket: New Feature",
    ],
    "commitSHA": "a0000003",
    "date": "2017-01-01",
    "gitHostingIssue": Object {
      "labels": Array [
        Object {
          "name": "Type: New Feature",
        },
        Object {
          "name": "Status: In Progress",
        },
      ],
      "number": 2,
      "title": "This is the commit title for the issue (#2)",
      "user": Object {
        "html_url": "https://github.com/test-user",
        "login": "test-user",
        "name": "Test User",
      },
    },
    "issueNumber": "2",
    "message": "feat(module) Add new module (#2)",
    "packages": Array [],
    "tags": undefined,
  },
  Object {
    "commitSHA": "a0000002",
    "date": "2017-01-01",
    "issueNumber": null,
    "message": "refactor(module) Simplify implementation",
    "packages": Array [],
    "tags": undefined,
  },
  Object {
    "commitSHA": "a0000001",
    "date": "2017-01-01",
    "issueNumber": null,
    "message": "chore(release): releasing component",
    "packages": Array [],
    "tags": Array [
      "v0.1.0",
    ],
  },
]
`;
